p11kit/p11kit.go

```diff
@@ -541,6 +544,44 @@ func (h *handler) handleGetMechanismInfo(req *body) (*body, error) {
 	return resp, nil
 }
 
+func (h *handler) handleGenerateRandom(req *body) (*body, error) {
+	// https://github.com/p11-glue/p11-kit/blob/0.24.0/p11-kit/rpc-client.c#L1843
+
+	// We don't actually want to support C_GenerateRandom, but some clients may
+	// blindly use it without checking the return value. This function still reports
+	// that we don't have a RNG, but fills the provided bytes with random data from
+	// /dev/urandom anyway in order to provide protection for a client that doesn't
+	// properly check the C_GenerateRandom return value.
+
+	var (
+		sessionID  uint64
+		randomData []byte
+		randomLen  uint32
+	)
+	req.readUlong(&sessionID)
+	req.readByteArray(&randomData, &randomLen)
+
+	if err := req.err(); err != nil {
+		return nil, err
+	}
+
+	_, err := rand.Read(randomData)
```

Comment:
This won't be sent if you send errRandomNoRNG. Is that what you want? If so you need to return a response object and not an error. If not this can be nuked

Reply:
Ah right.. Hmmm... In the old shim we populated the buffer but returned an error (since it was just C all the way down). Now that we have so many layers of indirection it *hopefully* makes sense to just return the error and pray the client doesn't ignore it.